pkg/eestream/decode.go

```diff
@@ -47,41 +80,86 @@ func (dr *decodedReader) Read(p []byte) (n int, err error) {
 		if dr.err != nil {
 			return 0, err
 		}
-		// we're going to kick off a bunch of goroutines. make a
-		// channel to catch those goroutine errors. importantly,
-		// the channel has a buffer size to contain all the errors
-		// even if we read none, so we can return without receiving
-		// every channel value
-		errs := make(chan readerError, len(dr.rs))
-		for i := range dr.rs {
-			go func(i int) {
-				// fill the buffer from the piece input
-				_, err := io.ReadFull(dr.rs[i], dr.inbufs[i])
-				errs <- readerError{i, err}
-			}(i)
-		}
-		// catch all the errors
-		inbufs := make(map[int][]byte, len(dr.inbufs))
+		// Run a selector on the readers' buffered channels
 		eofbufs := 0
-		for range dr.rs {
-			re := <-errs
-			if re.err == nil {
-				// add inbuf for decoding only if no error
-				inbufs[re.i] = dr.inbufs[re.i]
-			} else if re.err == io.EOF {
-				// keep track of inbufs at EOF
-				eofbufs++
+		inbufs := make(map[int][]byte, len(dr.chans))
+		// use reflect to select from the array of channels
+		cases := make([]reflect.SelectCase, len(dr.chans)+1)
+		// default case for non-blocking selection
+		cases[0] = reflect.SelectCase{
+			Dir: reflect.SelectDefault, Chan: reflect.Value{}}
+		for i, ch := range dr.chans {
+			cases[i+1] = reflect.SelectCase{
+				Dir: reflect.SelectRecv, Chan: reflect.ValueOf(ch)}
+		}
+		// iterate until the new block is received from enough channels
+		for len(cases) > 1 {
+			// non-blocking select - harvest available input buffers
+			chosen, value, ok := reflect.Select(cases)
+			if chosen == 0 {
+				// default case - no more input buffers available, check if we have enough
+				// TODO is required+1 enough to detect errors in polynomials of any degree?
+				if len(inbufs) >= dr.es.RequiredCount()+1 ||
+					len(inbufs) >= dr.es.TotalCount() {
+					// we have enough input buffers, fill the decoded output buffer
+					dr.outbuf, dr.err = dr.es.Decode(dr.outbuf, inbufs)
+					if dr.err == nil {
+						break
+					}
+					// TODO is there better way for error comparision?
```

Comment:
yeah so we should push a change upstream to infectious to have different error types. infectious uses spacemonkeygo/errors which is like zeebo/errs where there are error classes and you can check if an error is of a certain class. 

that said, it might be better to extend infectious to return something more than an error, maybe an infectious.Error which is an error but also has methods that return these values. that way the type checker will make sure we're checking the right thing

either way, i think infectious needs to change if we want to handle these errors specially